crates/bevy_render/src/render_resource/resource_macros.rs

```diff
@@ -1,90 +1,90 @@
 // structs containing wgpu types take a long time to compile. this is particularly bad for generic
 // structs containing wgpu structs. we avoid that in debug builds (and for cargo check and rust analyzer)
-// by boxing and type-erasing with the `render_resource_wrapper` macro.
+// by type-erasing with the `render_resource_wrapper` macro. The resulting type behaves like Arc<$wgpu_type>,
+// but avoids explicitly storing an Arc<$wgpu_type> member.
 // analysis from https://github.com/bevyengine/bevy/pull/5950#issuecomment-1243473071 indicates this is
 // due to `evaluate_obligations`. we should check if this can be removed after a fix lands for
 // https://github.com/rust-lang/rust/issues/99188 (and after other `evaluate_obligations`-related changes).
 #[cfg(debug_assertions)]
 #[macro_export]
 macro_rules! render_resource_wrapper {
     ($wrapper_type:ident, $wgpu_type:ty) => {
-        #[derive(Clone, Debug)]
-        pub struct $wrapper_type(Option<std::sync::Arc<Box<()>>>);
+        #[derive(Debug)]
+        // safety invariant: when self.0 is Some(ptr), it comes from `into_raw` of an Arc<$wgpu_type> with a strong ref.
+        pub struct $wrapper_type(Option<*const ()>);
```

Comment:
yes, so that when we consume the arc in try_unwrap we don't then consume it again in Drop

Reply:
Can we just `forget` the Arc in `try_unwrap`?

Reply:
i mentioned that on the original issue [here](https://github.com/bevyengine/bevy/pull/5950#discussion_r983317192). i can if you like, i think it would be functionally the same, but i feel it would make the safety invariant less clear to verify (that the arc is valid iff the pointer is Some). 

are you looking to avoid using extra space for the option?

Reply:
Both the extra space and the extra branch on all of these operations. If that's removed, we can probably use this implementation for the release mode too and avoid the need for a cfg block. (Provided it works out in the benchmarks/stress tests).

Not too big a deal if we avoid this. This will likely all go away once the upstream compiler issue is fixed.

Reply:
option removed. 

its not committed here but i've tested in release, i see no impact on many_foxes and the compile time for changing code in bevy_render drops by around 40% there as well. not sure what the best benchmark would be though. either way i guess that should probably be a separate pr.